Create Native client/server for Android server/client from BLE Service

ychernysheva opened the pull request:
Fixes [FORDTCN-12046](https://adc.luxoft.com/jira/browse/FORDTCN-12046)

ToDo: write detailed description?

VadymLuchko committed:
Pass some data from blessed to native

VadymLuchko committed:
connection to local socket

ychernysheva committed:
Process request and response from native side via unix sockets

AKalinich-Luxoft commented on app/src/main/cpp/src/components/transport_manager/include/transport_manager/bluetooth_le/bluetooth_le_connection_factory.h:
```diff
@@ -0,0 +1,98 @@
+  /**
+   * @brief Start BT connection factory.
+   */
+  virtual TransportAdapter::Error Init();
```
Please mark methods with `override` instead of adding virtual

ychernysheva replied:
Done in 1537bc4

AKalinich-Luxoft commented on app/src/main/cpp/src/components/transport_manager/src/bluetooth_le/ble_client.cc:
```diff
@@ -0,0 +1,75 @@
+
+void BleClient::Init()
+{
+    SDL_LOG_TRACE("enter");
```
@ychernysheva why not SDL_LOG_AUTO_TRACE ?

ychernysheva replied:
Done in 1537bc4

AKalinich-Luxoft commented on app/src/main/cpp/src/components/transport_manager/src/bluetooth_le/ble_client.cc:
```diff
@@ -0,0 +1,75 @@
+    if (socket_id < 0) {
+        err = errno;
+        SDL_LOG_INFO("Cannot open socket: " << strerror(err));
+        errno = err;
```
@ychernysheva what is the sense of this line?

ychernysheva replied:
Removed in 1537bc4

AKalinich-Luxoft commented on app/src/main/cpp/src/components/transport_manager/src/bluetooth_le/ble_client.cc:
```diff
@@ -0,0 +1,75 @@
+    socket_id = socket(PF_LOCAL, SOCK_STREAM, 0);
+    if (socket_id < 0) {
+        err = errno;
+        SDL_LOG_INFO("Cannot open socket: " << strerror(err));
```
```suggestion
        SDL_LOG_ERROR("Cannot open socket: " << strerror(err));
```

ychernysheva replied:
Done in 1537bc4

AKalinich-Luxoft commented on app/src/main/cpp/src/components/transport_manager/src/bluetooth_le/ble_client.cc:
```diff
@@ -0,0 +1,75 @@
+
+    if (connect(socket_id, (struct sockaddr *) &addr, len) < 0) {
+        err = errno;
+        SDL_LOG_INFO("connect() failed: " <<
```
```suggestion
        SDL_LOG_ERROR("connect() failed: " <<
```

ychernysheva replied:
Done in 1537bc4

AKalinich-Luxoft commented on app/src/main/cpp/src/components/transport_manager/src/bluetooth_le/ble_client.cc:
```diff
@@ -0,0 +1,75 @@
+        SDL_LOG_INFO("connect() failed: " <<
+             strerror(err));
+        close(socket_id);
+        errno = err;
```
@ychernysheva what is the sense of that?

ychernysheva replied:
Removed in 1537bc4

AKalinich-Luxoft commented on app/src/main/cpp/src/components/transport_manager/include/transport_manager/bluetooth_le/ble_client.h:
```diff
@@ -0,0 +1,25 @@
+
+class BleClient
+{
+    int socket_id;
+    bool connected;
```
@ychernysheva please use names according to our coding style
```suggestion
    int socket_id_;
    bool connected_;
```

ychernysheva replied:
Done in 1537bc4

AKalinich-Luxoft commented on app/src/main/cpp/src/components/transport_manager/src/bluetooth_le/ble_client.cc:
```diff
@@ -0,0 +1,75 @@
+
+void BleClient::Run()
+{
+    SDL_LOG_TRACE("enter");
```
@ychernysheva SDL_LOG_AUTO_TRACE

ychernysheva replied:
Done in 1537bc4

AKalinich-Luxoft commented on app/src/main/cpp/src/components/transport_manager/src/bluetooth_le/ble_client.cc:
```diff
@@ -0,0 +1,75 @@
+    {
+        close(socket_id);
+    }
+}
```
@ychernysheva add empty line here

ychernysheva replied:
Done in 1537bc4

AKalinich-Luxoft commented on app/src/main/cpp/src/components/transport_manager/src/bluetooth_le/ble_server.cc:
```diff
@@ -0,0 +1,92 @@
+    const char* socket_name =  "./localBleWriter";
+
+    char buffer[256];
+    bool thread_stop = false;
```
@ychernysheva what is the purpose of that variable?

ychernysheva replied:
For now I replace it with true to make this cycle infinite

AKalinich-Luxoft commented on app/src/main/cpp/src/components/transport_manager/src/bluetooth_le/ble_server.cc:
```diff
@@ -0,0 +1,92 @@
+namespace{
+    const char* socket_name =  "./localBleWriter";
+
+    char buffer[256];
```
@ychernysheva I would use for a separate constant for a buffer size here and at line 15

ychernysheva replied:
And I believe that buffer size needs to be equal MTU?

ychernysheva replied:
But i'm not sure how to transfer this data from Java side, if needed?

ychernysheva replied:
May I suggest to store MTU value in some external file, for example in ini file, to read this value from Java and Native side?

AKalinich-Luxoft replied:
@ychernysheva let's keep it simple for this PR. Just declare a const variable at the beginning of this file and use it. We will think about aligning with MTU later in a separate PR if needed.

ychernysheva replied:
Done in d331af8

AKalinich-Luxoft commented on app/src/main/cpp/src/components/transport_manager/include/transport_manager/bluetooth_le/ble_server.h:
```diff
@@ -0,0 +1,25 @@
+
+class BleServer
+{
+    int server_sock;
+    int client_sock;
+    bool connected;
```
```suggestion
    int server_sock_;
    int client_sock_;
    bool connected_;
```

ychernysheva replied:
Done in 1537bc4

AKalinich-Luxoft commented on app/src/main/cpp/src/components/transport_manager/src/bluetooth_le/ble_server.cc:
```diff
@@ -0,0 +1,92 @@
+    if (server_sock < 0) {
+        err = errno;
+        SDL_LOG_INFO("Cannot open socket: " << strerror(err));
+        errno = err;
```
@ychernysheva same as above

ychernysheva replied:
Done in 1537bc4

AKalinich-Luxoft commented on app/src/main/cpp/src/components/transport_manager/src/bluetooth_le/ble_server.cc:
```diff
@@ -0,0 +1,92 @@
+    server_sock = socket(PF_LOCAL, SOCK_STREAM, 0);
+    if (server_sock < 0) {
+        err = errno;
+        SDL_LOG_INFO("Cannot open socket: " << strerror(err));
```
```suggestion
        SDL_LOG_ERROR("Cannot open socket: " << strerror(err));
```

ychernysheva replied:
Done in 1537bc4

AKalinich-Luxoft commented on app/src/main/java/org/luxoft/sdl_core/BleLocalSocketReader.java:
```diff
@@ -98,7 +102,13 @@ public void run() {
                             + stringified_data);
-                    mCallback.OnMessageReceived(buffer);
+                    String new_string = new String("message " + counter);
+                    counter += 1;
```
```suggestion
                    ++counter;
```

ychernysheva replied:
Done in 1537bc4

AKalinich-Luxoft commented on app/src/main/java/org/luxoft/sdl_core/BleLocalSocketReader.java:
```diff
@@ -98,7 +102,13 @@ public void run() {
+                    String new_string = new String("message " + counter);
+                    counter += 1;
+
+                    byte[] new_buffer;
+                    new_buffer = new_string.getBytes();
```
Is it possible to merge these two lines into a single one?

ychernysheva replied:
done in 1537bc4

AKalinich-Luxoft commented on app/src/main/java/org/luxoft/sdl_core/BleLocalSocketReader.java:
```diff
@@ -98,7 +102,13 @@ public void run() {
+
+                    byte[] new_buffer;
+                    new_buffer = new_string.getBytes();
+                    //mCallback.OnMessageReceived(buffer);
```
Please remove this comment

ychernysheva replied:
Done in 1537bc4

ychernysheva committed:
Fix synchronisation issues

ychernysheva committed:
fixup! Fix synchronisation issues

AKalinich-Luxoft committed:
Make server init before client init

AKalinich-Luxoft committed:
Add reconnect feature for Java/Native clients

We don't have any guarantees regarding the timings and when
Java/Native server become available for accepting socket connetctions.
By that reason, if client is attempting to connect to early, connection
might be refused. To avoid this, client should wait for some time and
try to reconnect again several attempts.

AKalinich-Luxoft committed:
Move BLE client and server to transport namespace

AKalinich-Luxoft committed:
Add possibility to send RawMessage from client

Was added MessageQueue and updated Run() lifecycle to work with it
instead of hardcoded logic.

AKalinich-Luxoft committed:
Add possibility to forward RawMessage from BLE server

Also updated Run() lifecycle to work with raw messages and
interract with transport adapter

AKalinich-Luxoft committed:
fixup! Add reconnect feature for Java/Native clients

AKalinich-Luxoft committed:
Add test function for sending responses to Java part

AKalinich-Luxoft committed:
Start Java server thread upon Initialization

It is important to keep two way communication even if there is
no any BLE device connected.

AKalinich-Luxoft committed:
Disconnect reader before writer to keep order

AKalinich-Luxoft committed:
Synchronize callback and fix array forwarded size

AKalinich-Luxoft committed:
Fix application crash on service restart

Broadcast receiver registration/unregistration should be done on
service create/destroy but not on its start. It causes crash on attempt
to start it one more time.

AKalinich-Luxoft committed:
Start and stop adapter thread properly

It should be stopped on stop BLE routine and recreated on attempt
to start BLE routine again.

AKalinich-Luxoft committed:
Add test timer to emulate messages from mobile

TODO: remove this code later
